source/opt/loop_unroller.cpp

```diff
@@ -0,0 +1,793 @@
+
+  // Copy a given |block_to_copy| in the |loop| and record the mapping of the
+  // old/new ids. |preserve_instructions| determines whether or not the method
+  // will modify (other than result_id) instructions which are copied.
+  void CopyBasicBlock(ir::Loop* loop, const ir::BasicBlock* block_to_copy,
+                      bool preserve_instructions);
+
+  // The actual implementation of the unroll step. Unrolls |loop| by given
+  // |factor| by copying the body by |factor| times. Also propagates the
+  // induction variable value throughout the copies.
+  void Unroll(ir::Loop* loop, size_t);
+
+  // Fills the loop_blocks_inorder_ field with the ordered list of basic blocks
+  // as computed by the method ComputeLoopOrderedBlocks.
+  void ComputeLoopOrderedBlocks(ir::Loop* loop);
+
+  // A pointer to the IRContext. Used to add/remove instructions and for usedef
+  // chains.
+  ir::IRContext* ir_context_;
+
+  // A reference the function the loop is within.
+  ir::Function& function_;
+
+  // A list of basic blocks to be added to the loop at the end of an unroll
+  // step.
+  BasicBlockListTy blocks_to_add_;
+
+  // Maintains the current state of the transform between calls to unroll.
+  LoopUnrollState state_;
+
+  // An ordered list containing the loop basic blocks.
+  std::vector<ir::BasicBlock*> loop_blocks_inorder_;
+
+  // The block containing the condition check which contains a conditional
+  // branch to the merge and continue block.
+  ir::BasicBlock* loop_condition_block_;
+
+  // The induction variable of the loop.
+  ir::Instruction* loop_induction_variable_;
+
+  // The number of loop iterations that the loop would preform pre-unroll.
+  size_t number_of_loop_iterations_;
+};
+
+/*
+ * Static helper functions.
+ */
+
+// Retrieve the index of the OpPhi instruction |phi| which corresponds to the
+// incoming |block| id.
+static uint32_t GetPhiIndexFromLabel(const ir::BasicBlock* block,
+                                     const ir::Instruction* phi) {
+  for (uint32_t i = 1; i < phi->NumInOperands(); i += 2) {
+    if (block->id() == phi->GetSingleWordInOperand(i)) {
+      return i;
+    }
+  }
+  assert(false && "Could not find operand in instruction.");
+  return 0;
+}
+
+void LoopUnrollerUtilsImpl::Init(ir::Loop* loop) {
+  loop_condition_block_ = loop->FindConditionBlock(function_);
+
+  // When we reinit the second loop during PartiallyUnrollUnevenFactor we need
+  // to use the cached value from the duplicate step as the dominator tree
+  // basded solution, loop->FindConditionBlock, requires all the nodes to be
+  // connected up with the correct branches. They won't be at this point.
+  if (!loop_condition_block_) {
+    loop_condition_block_ = state_.new_condition_block;
+  }
+  assert(loop_condition_block_);
+
+  loop_induction_variable_ = loop->FindInductionVariable(loop_condition_block_);
+  assert(loop_induction_variable_);
+
+#ifndef NDEBUG
+  bool found =
+#endif
```

Comment:
I'm assuming you did this because of the unused variable error for a release build.  I think you can add a line:
```
bool found = ...;
found;  // So the release build does not complain.
assert(found);
```
I think this works and it is a little cleaner.

Reply:
fwiw `(void)found;` is a canonical way to suppress unused variable warning; `found;` causes warnings on some compilers because it's an unused expression